src/extensions/lock_time/types.rs

```diff
@@ -0,0 +1,205 @@
+use crate::extensions::MATURE_THRESHOLD;
+
+use ckb_indexer::store;
+use ckb_types::core::{BlockNumber, RationalU256};
+use ckb_types::{packed, prelude::*};
+use derive_more::Display;
+use serde::{Deserialize, Serialize};
+
+use std::collections::VecDeque;
+
+#[derive(Debug, Display)]
+pub enum LocktimeExtensionError {
+    #[display(fmt = "DB Error {}", _0)]
+    DBError(String),
+}
+
+impl std::error::Error for LocktimeExtensionError {}
+
+impl From<store::Error> for LocktimeExtensionError {
+    fn from(err: store::Error) -> LocktimeExtensionError {
+        LocktimeExtensionError::DBError(err.to_string())
+    }
+}
+
+#[repr(u8)]
+pub enum KeyPrefix {
+    Address = 0,
+    Block = 16,
+}
+
+#[derive(Clone, Debug)]
+pub enum Key<'a> {
+    CkbAddress(&'a str),
+    Block(BlockNumber, &'a packed::Byte32),
+}
+
+impl<'a> Into<Vec<u8>> for Key<'a> {
+    fn into(self) -> Vec<u8> {
+        let mut encoded = Vec::new();
+
+        match self {
+            Key::CkbAddress(key) => {
+                encoded.push(KeyPrefix::Address as u8);
+                encoded.push(key.as_bytes().len() as u8);
+                encoded.extend_from_slice(key.as_bytes());
+            }
+
+            Key::Block(block_num, block_hash) => {
+                encoded.push(KeyPrefix::Block as u8);
+                encoded.extend_from_slice(&block_num.to_be_bytes());
+                encoded.extend_from_slice(block_hash.as_slice());
+            }
+        }
+
+        encoded
+    }
+}
+
+impl<'a> Key<'a> {
+    pub fn into_vec(self) -> Vec<u8> {
+        self.into()
+    }
+}
+
+#[derive(Clone, Debug)]
+pub enum Value {
+    CellbaseCapacity(Vec<u8>),
+    RollbackData(Vec<u8>),
+}
+
+impl Into<Vec<u8>> for Value {
+    fn into(self) -> Vec<u8> {
+        match self {
+            Value::CellbaseCapacity(data) => data,
+            Value::RollbackData(data) => data,
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Default, Clone, Debug, PartialEq, Eq)]
+pub struct CellbaseCkbAccount {
+    maturity: u64,
+    immature: VecDeque<CellbaseCkb>,
+}
+
+impl CellbaseCkbAccount {
+    pub fn push(&mut self, cellbase: CellbaseCkb) {
+        self.immature.push_back(cellbase);
+    }
+
+    pub fn remove(&mut self, cellbase: &CellbaseCkb) {
+        let mut index = usize::MAX;
+
+        for (idx, item) in self.immature.iter().enumerate() {
+            if item == cellbase {
+                index = idx;
+                break;
+            }
+        }
+
+        if index != usize::MAX {
+            self.immature.remove(index);
+        }
+    }
+
+    pub fn mature(&mut self) {
+        let mut mature_ckb = 0u64;
+        let threshold = MATURE_THRESHOLD.read();
```

Comment:
I think `Threshold = current_epoch + mature_threshold`.  At least you should pass the `current_epoch`